src/test/rgw/test_d4n_filter.cc

```diff
@@ -2759,6 +2952,437 @@ TEST_F(D4NFilterFixture, CopyReplaceVersionedObjectWrite)
+  }, rethrow);
+
+  io.run_for(std::chrono::seconds(2)); // Allow cleaning cycle to complete
+
+  net::spawn(io, [this, &testName, &bucketName, &deleteMarker] (net::yield_context yield) {
+    dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + deleteMarker, err), false);  
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.reset();
+  io.run();
+}
+
+TEST_F(D4NFilterFixture, VersionedDeleteBeforeCleaning)
+{
+  env->cct->_conf->d4n_writecache_enabled = true;
+  env->cct->_conf->rgw_d4n_cache_cleaning_interval = 0.5;
+  const std::string testName = "VersionedDeleteBeforeCleaning";
+  const std::string bucketName = "/tmp/d4n_filter_tests/dbstore-default_ns.1";
+  std::vector<std::string> instances;
+ 
+  net::spawn(io, [this, &testName, &bucketName, &instances] (net::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    testBucket->get_info().bucket.bucket_id = bucketName;
+    
+    for (int i = 0; i <= 1; ++i) {
+      std::string instance;
+      put_version_enabled_object(testName, instance, yield); // Upload two versions 
+      instances.push_back(instance);
+    }
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("ZREVRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "-1");
+
+      response< std::vector<std::string> > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ(std::get<0>(resp).value().size(), 2);
+      EXPECT_EQ(std::get<0>(resp).value()[0], instances[1]);
+      EXPECT_EQ(std::get<0>(resp).value()[1], instances[0]);
+    }
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + instances[1], err), true);  
+    std::string oid = instances[1] + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + url_encode(bucketName, true) + "/" + TEST_OBJ + testName + "/" + oid, err), true);  
+
+    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op_enabled = objEnabled->get_delete_op();
+    objEnabled->set_instance(instances[1]); // Latest version
+    EXPECT_EQ(del_op_enabled->delete_obj(env->dpp, optional_yield{yield}, rgw::sal::FLAG_LOG_OP), 0);
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("ZREVRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "-1");
+
+      response< std::vector<std::string> > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ(std::get<0>(resp).value().size(), 1);
+      EXPECT_EQ(std::get<0>(resp).value()[0], instances[0]);
+    }
+
```

Comment:
along with these checks you also need to perform a simple 'get' and confirm whether the versionId returned is the correct next latest version.